File: api_examples.js
/*
 * Django IoT Dashboard - API Examples (JavaScript)
 * ตัวอย่างการใช้งาน REST API ด้วย JavaScript fetch
 */

// ========================================
// Configuration
// ========================================
const BASE_URL = 'http://localhost:8000';
const API_V1 = `${BASE_URL}/api/v1`;

// ========================================
// Helper Functions
// ========================================

async function apiRequest(url, method = 'GET', body = null) {
    const options = {
        method: method,
        headers: {
            'Content-Type': 'application/json'
        }
    };
    
    if (body) {
        options.body = JSON.stringify(body);
    }
    
    try {
        const response = await fetch(url, options);
        const data = await response.json();
        return data;
    } catch (error) {
        return {
            success: false,
            error: error.message
        };
    }
}

// ========================================
// LED Control Functions
// ========================================

async function getLedStatus() {
    return await apiRequest(`${API_V1}/led/`);
}

async function controlLed(command) {
    // command: 'ON', 'OFF', or 'TOGGLE'
    return await apiRequest(`${API_V1}/led/`, 'POST', { command });
}

// ========================================
// Relay Control Functions
// ========================================

async function getRelayStatus() {
    return await apiRequest(`${API_V1}/relay/`);
}

async function controlRelay(relay_num, command) {
    // relay_num: 1, 2, or 3
    // command: 'ON', 'OFF', or 'TOGGLE'
    return await apiRequest(`${API_V1}/relay/`, 'POST', {
        relay_num,
        command
    });
}

// ========================================
// Sensor Data Functions
// ========================================

async function getLatestSensor() {
    return await apiRequest(`${API_V1}/sensors/latest/`);
}

async function getSensorList(limit = 20, offset = 0) {
    return await apiRequest(`${API_V1}/sensors/?limit=${limit}&offset=${offset}`);
}

async function getSensorStats() {
    return await apiRequest(`${API_V1}/sensors/stats/`);
}

async function createSensorData(temperature, humidity, device_name = 'JS_API_Test') {
    return await apiRequest(`${API_V1}/sensors/`, 'POST', {
        temperature,
        humidity,
        device_name
    });
}

async function getSensorById(sensor_id) {
    return await apiRequest(`${API_V1}/sensors/${sensor_id}/`);
}

async function updateSensorData(sensor_id, data) {
    return await apiRequest(`${API_V1}/sensors/${sensor_id}/`, 'PATCH', data);
}

async function deleteSensorData(sensor_id) {
    return await apiRequest(`${API_V1}/sensors/${sensor_id}/`, 'DELETE');
}

// ========================================
// System Status Functions
// ========================================

async function getSystemStatus() {
    return await apiRequest(`${API_V1}/system/status/`);
}

// ========================================
// Example Usage
// ========================================

async function runExamples() {
    console.log('='.repeat(60));
    console.log('Django IoT Dashboard - API Examples (JavaScript)');
    console.log('='.repeat(60));
    console.log('');
    
    // Example 1: Get System Status
    console.log('📊 Example 1: Get System Status');
    console.log('-'.repeat(60));
    let result = await getSystemStatus();
    if (result.success) {
        console.log(JSON.stringify(result.data, null, 2));
    } else {
        console.log(`Error: ${result.error}`);
    }
    console.log('');
    
    // Example 2: Control LED
    console.log('💡 Example 2: Control LED');
    console.log('-'.repeat(60));
    
    // Turn ON
    console.log('→ Turning LED ON...');
    result = await controlLed('ON');
    console.log(`  Status: ${result.message || result.error}`);
    await sleep(1000);
    
    // Check status
    result = await getLedStatus();
    if (result.success) {
        console.log(`  LED is now: ${result.data.is_on ? 'ON 🟢' : 'OFF ⚫'}`);
    }
    await sleep(2000);
    
    // Turn OFF
    console.log('→ Turning LED OFF...');
    result = await controlLed('OFF');
    console.log(`  Status: ${result.message || result.error}`);
    console.log('');
    
    // Example 3: Control Relays
    console.log('⚡ Example 3: Control Relays');
    console.log('-'.repeat(60));
    
    // Turn ON Relay 1
    console.log('→ Turning Relay 1 ON...');
    result = await controlRelay(1, 'ON');
    console.log(`  Status: ${result.message || result.error}`);
    await sleep(1000);
    
    // Turn ON Relay 2
    console.log('→ Turning Relay 2 ON...');
    result = await controlRelay(2, 'ON');
    console.log(`  Status: ${result.message || result.error}`);
    await sleep(1000);
    
    // Check all relays
    result = await getRelayStatus();
    if (result.success) {
        const data = result.data;
        console.log(`  Relay 1: ${data.relay1 ? 'ON 🟢' : 'OFF ⚫'}`);
        console.log(`  Relay 2: ${data.relay2 ? 'ON 🟢' : 'OFF ⚫'}`);
        console.log(`  Relay 3: ${data.relay3 ? 'ON 🟢' : 'OFF ⚫'}`);
    }
    await sleep(1000);
    
    // Turn OFF all relays
    console.log('→ Turning all relays OFF...');
    await controlRelay(1, 'OFF');
    await controlRelay(2, 'OFF');
    await controlRelay(3, 'OFF');
    console.log('  All relays turned OFF');
    console.log('');
    
    // Example 4: Sensor Data
    console.log('🌡️  Example 4: Sensor Data Management');
    console.log('-'.repeat(60));
    
    // Get latest sensor
    console.log('→ Getting latest sensor data...');
    result = await getLatestSensor();
    if (result.success && result.data) {
        const data = result.data;
        console.log(`  Temperature: ${data.temperature}°C`);
        console.log(`  Humidity: ${data.humidity}%`);
        console.log(`  Time: ${data.timestamp}`);
    } else {
        console.log('  No sensor data available');
    }
    console.log('');
    
    // Create test sensor data
    console.log('→ Creating test sensor data...');
    result = await createSensorData(28.5, 65.3, 'JS_API_Test');
    if (result.success) {
        const sensor_id = result.data.id;
        console.log(`  Created sensor ID: ${sensor_id}`);
        console.log(`  Temperature: ${result.data.temperature}°C`);
        console.log(`  Humidity: ${result.data.humidity}%`);
        
        // Update the sensor data
        await sleep(1000);
        console.log(`→ Updating sensor ${sensor_id}...`);
        result = await updateSensorData(sensor_id, { temperature: 29.0 });
        if (result.success) {
            console.log(`  Updated temperature to: ${result.data.temperature}°C`);
        }
        
        // Delete the sensor data
        await sleep(1000);
        console.log(`→ Deleting sensor ${sensor_id}...`);
        result = await deleteSensorData(sensor_id);
        if (result.success) {
            console.log('  Deleted successfully');
        }
    }
    console.log('');
    
    // Example 5: Get Statistics
    console.log('📈 Example 5: Sensor Statistics');
    console.log('-'.repeat(60));
    result = await getSensorStats();
    if (result.success) {
        const stats = result.data;
        console.log(`  Total Readings: ${stats.total_readings}`);
        console.log('  Temperature:');
        console.log(`    - Average: ${stats.temperature.average}°C`);
        console.log(`    - Max: ${stats.temperature.max}°C`);
        console.log(`    - Min: ${stats.temperature.min}°C`);
        console.log('  Humidity:');
        console.log(`    - Average: ${stats.humidity.average}%`);
        console.log(`    - Max: ${stats.humidity.max}%`);
        console.log(`    - Min: ${stats.humidity.min}%`);
    }
    console.log('');
    
    // Example 6: Get Sensor List with Pagination
    console.log('📋 Example 6: Get Sensor List (Pagination)');
    console.log('-'.repeat(60));
    result = await getSensorList(5, 0);
    if (result.success) {
        console.log(`  Total: ${result.pagination.total}`);
        console.log(`  Showing: ${result.pagination.count} items`);
        console.log('  Latest 5 readings:');
        result.data.forEach(sensor => {
            console.log(`    - ID ${sensor.id}: ${sensor.temperature}°C, ${sensor.humidity}% @ ${sensor.timestamp}`);
        });
    }
    console.log('');
    
    console.log('='.repeat(60));
    console.log('✅ All examples completed!');
    console.log('='.repeat(60));
}

// Helper sleep function
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// ========================================
// Browser Usage (add to HTML)
// ========================================

/*
<!DOCTYPE html>
<html>
<head>
    <title>Django IoT Dashboard API Test</title>
</head>
<body>
    <h1>Django IoT Dashboard API Test</h1>
    
    <div>
        <h2>LED Control</h2>
        <button onclick="controlLed('ON')">Turn ON</button>
        <button onclick="controlLed('OFF')">Turn OFF</button>
        <button onclick="controlLed('TOGGLE')">Toggle</button>
        <button onclick="showLedStatus()">Get Status</button>
        <div id="led-status"></div>
    </div>
    
    <div>
        <h2>Relay Control</h2>
        <div>
            <h3>Relay 1</h3>
            <button onclick="controlRelay(1, 'ON')">ON</button>
            <button onclick="controlRelay(1, 'OFF')">OFF</button>
            <button onclick="controlRelay(1, 'TOGGLE')">Toggle</button>
        </div>
        <div>
            <h3>Relay 2</h3>
            <button onclick="controlRelay(2, 'ON')">ON</button>
            <button onclick="controlRelay(2, 'OFF')">OFF</button>
            <button onclick="controlRelay(2, 'TOGGLE')">Toggle</button>
        </div>
        <div>
            <h3>Relay 3</h3>
            <button onclick="controlRelay(3, 'ON')">ON</button>
            <button onclick="controlRelay(3, 'OFF')">OFF</button>
            <button onclick="controlRelay(3, 'TOGGLE')">Toggle</button>
        </div>
        <button onclick="showRelayStatus()">Get All Status</button>
        <div id="relay-status"></div>
    </div>
    
    <div>
        <h2>Sensor Data</h2>
        <button onclick="showLatestSensor()">Get Latest</button>
        <button onclick="showSensorStats()">Get Stats</button>
        <button onclick="createTestSensor()">Create Test Data</button>
        <div id="sensor-data"></div>
    </div>
    
    <div>
        <h2>System Status</h2>
        <button onclick="showSystemStatus()">Get System Status</button>
        <div id="system-status"></div>
    </div>
    
    <script src="api_examples.js"></script>
    <script>
        async function showLedStatus() {
            const result = await getLedStatus();
            document.getElementById('led-status').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
        
        async function showRelayStatus() {
            const result = await getRelayStatus();
            document.getElementById('relay-status').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
        
        async function showLatestSensor() {
            const result = await getLatestSensor();
            document.getElementById('sensor-data').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
        
        async function showSensorStats() {
            const result = await getSensorStats();
            document.getElementById('sensor-data').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
        
        async function createTestSensor() {
            const result = await createSensorData(28.5, 65.3, 'Browser_Test');
            document.getElementById('sensor-data').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
        
        async function showSystemStatus() {
            const result = await getSystemStatus();
            document.getElementById('system-status').innerHTML = 
                `<pre>${JSON.stringify(result, null, 2)}</pre>`;
        }
    </script>
</body>
</html>
*/

// ========================================
// Node.js Usage
// ========================================

// For Node.js, install node-fetch first:
// npm install node-fetch
// 
// Then add at the top:
// const fetch = require('node-fetch');
//
// Run with:
// node api_examples.js

// Export functions for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        getLedStatus,
        controlLed,
        getRelayStatus,
        controlRelay,
        getLatestSensor,
        getSensorList,
        getSensorStats,
        createSensorData,
        getSensorById,
        updateSensorData,
        deleteSensorData,
        getSystemStatus,
        runExamples
    };
}

File: api_examples.py
"""
Django IoT Dashboard - API Examples (Python)
ตัวอย่างการใช้งาน REST API ด้วย Python requests
"""

import requests
import json
import time

# ========================================
# Configuration
# ========================================
BASE_URL = "http://localhost:8000"
API_V1 = f"{BASE_URL}/api/v1"


# ========================================
# LED Control Functions
# ========================================

def get_led_status():
    """Get current LED status"""
    try:
        response = requests.get(f"{API_V1}/led/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def control_led(command):
    """
    Control LED
    command: 'ON', 'OFF', or 'TOGGLE'
    """
    try:
        response = requests.post(
            f"{API_V1}/led/",
            json={"command": command},
            headers={"Content-Type": "application/json"}
        )
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


# ========================================
# Relay Control Functions
# ========================================

def get_relay_status():
    """Get status of all relays"""
    try:
        response = requests.get(f"{API_V1}/relay/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def control_relay(relay_num, command):
    """
    Control specific relay
    relay_num: 1, 2, or 3
    command: 'ON', 'OFF', or 'TOGGLE'
    """
    try:
        response = requests.post(
            f"{API_V1}/relay/",
            json={
                "relay_num": relay_num,
                "command": command
            },
            headers={"Content-Type": "application/json"}
        )
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


# ========================================
# Sensor Data Functions
# ========================================

def get_latest_sensor():
    """Get latest sensor reading"""
    try:
        response = requests.get(f"{API_V1}/sensors/latest/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def get_sensor_list(limit=20, offset=0):
    """
    Get sensor data list with pagination
    limit: number of items per page
    offset: starting position
    """
    try:
        response = requests.get(
            f"{API_V1}/sensors/",
            params={"limit": limit, "offset": offset}
        )
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def get_sensor_stats():
    """Get sensor data statistics"""
    try:
        response = requests.get(f"{API_V1}/sensors/stats/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def create_sensor_data(temperature, humidity, device_name="API_Test"):
    """Create new sensor data entry"""
    try:
        response = requests.post(
            f"{API_V1}/sensors/",
            json={
                "temperature": temperature,
                "humidity": humidity,
                "device_name": device_name
            },
            headers={"Content-Type": "application/json"}
        )
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def get_sensor_by_id(sensor_id):
    """Get specific sensor data by ID"""
    try:
        response = requests.get(f"{API_V1}/sensors/{sensor_id}/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def update_sensor_data(sensor_id, temperature=None, humidity=None, device_name=None):
    """Update sensor data (partial update)"""
    try:
        data = {}
        if temperature is not None:
            data["temperature"] = temperature
        if humidity is not None:
            data["humidity"] = humidity
        if device_name is not None:
            data["device_name"] = device_name
        
        response = requests.patch(
            f"{API_V1}/sensors/{sensor_id}/",
            json=data,
            headers={"Content-Type": "application/json"}
        )
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


def delete_sensor_data(sensor_id):
    """Delete sensor data by ID"""
    try:
        response = requests.delete(f"{API_V1}/sensors/{sensor_id}/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


# ========================================
# System Status Functions
# ========================================

def get_system_status():
    """Get complete system status"""
    try:
        response = requests.get(f"{API_V1}/system/status/")
        return response.json()
    except Exception as e:
        return {"success": False, "error": str(e)}


# ========================================
# Example Usage
# ========================================

def main():
    print("=" * 60)
    print("Django IoT Dashboard - API Examples")
    print("=" * 60)
    print()
    
    # Example 1: Get System Status
    print("📊 Example 1: Get System Status")
    print("-" * 60)
    result = get_system_status()
    if result.get('success'):
        print(json.dumps(result['data'], indent=2))
    else:
        print(f"Error: {result.get('error')}")
    print()
    
    # Example 2: Control LED
    print("💡 Example 2: Control LED")
    print("-" * 60)
    
    # Turn ON
    print("→ Turning LED ON...")
    result = control_led('ON')
    print(f"  Status: {result.get('message', result.get('error'))}")
    time.sleep(1)
    
    # Check status
    result = get_led_status()
    if result.get('success'):
        print(f"  LED is now: {'ON 🟢' if result['data']['is_on'] else 'OFF ⚫'}")
    time.sleep(2)
    
    # Turn OFF
    print("→ Turning LED OFF...")
    result = control_led('OFF')
    print(f"  Status: {result.get('message', result.get('error'))}")
    print()
    
    # Example 3: Control Relays
    print("⚡ Example 3: Control Relays")
    print("-" * 60)
    
    # Turn ON Relay 1
    print("→ Turning Relay 1 ON...")
    result = control_relay(1, 'ON')
    print(f"  Status: {result.get('message', result.get('error'))}")
    time.sleep(1)
    
    # Turn ON Relay 2
    print("→ Turning Relay 2 ON...")
    result = control_relay(2, 'ON')
    print(f"  Status: {result.get('message', result.get('error'))}")
    time.sleep(1)
    
    # Check all relays
    result = get_relay_status()
    if result.get('success'):
        data = result['data']
        print(f"  Relay 1: {'ON 🟢' if data['relay1'] else 'OFF ⚫'}")
        print(f"  Relay 2: {'ON 🟢' if data['relay2'] else 'OFF ⚫'}")
        print(f"  Relay 3: {'ON 🟢' if data['relay3'] else 'OFF ⚫'}")
    time.sleep(1)
    
    # Turn OFF all relays
    print("→ Turning all relays OFF...")
    control_relay(1, 'OFF')
    control_relay(2, 'OFF')
    control_relay(3, 'OFF')
    print("  All relays turned OFF")
    print()
    
    # Example 4: Sensor Data
    print("🌡️  Example 4: Sensor Data Management")
    print("-" * 60)
    
    # Get latest sensor
    print("→ Getting latest sensor data...")
    result = get_latest_sensor()
    if result.get('success') and result['data']:
        data = result['data']
        print(f"  Temperature: {data['temperature']}°C")
        print(f"  Humidity: {data['humidity']}%")
        print(f"  Time: {data['timestamp']}")
    else:
        print("  No sensor data available")
    print()
    
    # Create test sensor data
    print("→ Creating test sensor data...")
    result = create_sensor_data(
        temperature=28.5,
        humidity=65.3,
        device_name="Python_API_Test"
    )
    if result.get('success'):
        sensor_id = result['data']['id']
        print(f"  Created sensor ID: {sensor_id}")
        print(f"  Temperature: {result['data']['temperature']}°C")
        print(f"  Humidity: {result['data']['humidity']}%")
        
        # Update the sensor data
        time.sleep(1)
        print(f"→ Updating sensor {sensor_id}...")
        result = update_sensor_data(sensor_id, temperature=29.0)
        if result.get('success'):
            print(f"  Updated temperature to: {result['data']['temperature']}°C")
        
        # Delete the sensor data
        time.sleep(1)
        print(f"→ Deleting sensor {sensor_id}...")
        result = delete_sensor_data(sensor_id)
        if result.get('success'):
            print(f"  Deleted successfully")
    print()
    
    # Example 5: Get Statistics
    print("📈 Example 5: Sensor Statistics")
    print("-" * 60)
    result = get_sensor_stats()
    if result.get('success'):
        stats = result['data']
        print(f"  Total Readings: {stats['total_readings']}")
        print(f"  Temperature:")
        print(f"    - Average: {stats['temperature']['average']}°C")
        print(f"    - Max: {stats['temperature']['max']}°C")
        print(f"    - Min: {stats['temperature']['min']}°C")
        print(f"  Humidity:")
        print(f"    - Average: {stats['humidity']['average']}%")
        print(f"    - Max: {stats['humidity']['max']}%")
        print(f"    - Min: {stats['humidity']['min']}%")
    print()
    
    # Example 6: Get Sensor List with Pagination
    print("📋 Example 6: Get Sensor List (Pagination)")
    print("-" * 60)
    result = get_sensor_list(limit=5, offset=0)
    if result.get('success'):
        print(f"  Total: {result['pagination']['total']}")
        print(f"  Showing: {result['pagination']['count']} items")
        print(f"  Latest 5 readings:")
        for sensor in result['data']:
            print(f"    - ID {sensor['id']}: {sensor['temperature']}°C, {sensor['humidity']}% @ {sensor['timestamp']}")
    print()
    
    print("=" * 60)
    print("✅ All examples completed!")
    print("=" * 60)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n❌ Interrupted by user")
    except Exception as e:
        print(f"\n\n❌ Error: {e}")
